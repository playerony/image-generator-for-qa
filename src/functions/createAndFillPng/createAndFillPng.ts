import { PNG } from 'pngjs/browser';
import type { CreateAndFillPngParams } from './createAndFillPng.types';
import { CHUNK_COLORS } from '~/constants';

export const createAndFillPng = ({ columnChunkAmount, parsedHeight, parsedWidth, rowChunkAmount, singleChunkHeight, singleChunkWidth }: CreateAndFillPngParams) => {
  const png = new PNG({ width: parsedWidth, height: parsedHeight });

  for (let i = 0; i < singleChunkWidth * singleChunkHeight; i++) {
    for (let j = 0; j < columnChunkAmount; j++) {
      const x = j * singleChunkWidth + (i % singleChunkWidth);

      for (let k = 0; k < rowChunkAmount; k++) {
        const y = k * singleChunkHeight + Math.floor(i / singleChunkWidth);
        const index = (x + y * parsedWidth) * 4;
        const { a, b, g, r } = CHUNK_COLORS[i % CHUNK_COLORS.length];

        png.data[index] = r;
        png.data[index + 1] = g;
        png.data[index + 2] = b;
        png.data[index + 3] = a;
      }
    }
  }

  return png;
};
