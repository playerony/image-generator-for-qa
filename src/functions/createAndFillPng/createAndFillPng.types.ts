export interface CreateAndFillPngParams {
  columnChunkAmount: number;
  parsedWidth: number;
  parsedHeight: number;
  rowChunkAmount: number;
  singleChunkHeight: number;
  singleChunkWidth: number;
}
