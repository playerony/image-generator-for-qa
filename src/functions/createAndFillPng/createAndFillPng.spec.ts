import { createAndFillPng } from './createAndFillPng';
import { vi, describe, it } from 'vitest';

vi.mock('pngjs/browser', () => ({
  PNG: vi.fn().mockImplementation(({ height, width }) => ({
    width,
    height,
    data: new Uint8Array(width * height * 4),
  })),
}));

vi.mock('~/constants', () => ({
  CHUNK_COLORS: [
    { r: 50, g: 100, b: 150, a: 255 },
    { r: 60, g: 120, b: 180, a: 255 },
    { r: 70, g: 140, b: 210, a: 255 },
  ],
}));

describe('createAndFillPng', () => {
  it('should fill the PNG data with the correct colors', ({ expect }) => {
    const params = {
      columnChunkAmount: 2,
      parsedHeight: 4,
      parsedWidth: 4,
      rowChunkAmount: 2,
      singleChunkHeight: 2,
      singleChunkWidth: 2,
    };

    const png = createAndFillPng(params);

    expect(png.data).toMatchSnapshot();
  });

  it('should create a PNG with the correct dimensions', ({ expect }) => {
    const params = {
      columnChunkAmount: 1,
      parsedHeight: 4,
      parsedWidth: 4,
      rowChunkAmount: 1,
      singleChunkHeight: 2,
      singleChunkWidth: 2,
    };

    const png = createAndFillPng(params);

    expect(png.data).toMatchSnapshot();
  });
});
