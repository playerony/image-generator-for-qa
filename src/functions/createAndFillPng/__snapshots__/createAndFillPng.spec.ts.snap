// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`createAndFillPng > should create a PNG with the correct dimensions 1`] = `
Uint8Array [
  50,
  100,
  150,
  255,
  60,
  120,
  180,
  255,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  70,
  140,
  210,
  255,
  50,
  100,
  150,
  255,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
]
`;

exports[`createAndFillPng > should fill the PNG data with the correct colors 1`] = `
Uint8Array [
  50,
  100,
  150,
  255,
  60,
  120,
  180,
  255,
  50,
  100,
  150,
  255,
  60,
  120,
  180,
  255,
  70,
  140,
  210,
  255,
  50,
  100,
  150,
  255,
  70,
  140,
  210,
  255,
  50,
  100,
  150,
  255,
  50,
  100,
  150,
  255,
  60,
  120,
  180,
  255,
  50,
  100,
  150,
  255,
  60,
  120,
  180,
  255,
  70,
  140,
  210,
  255,
  50,
  100,
  150,
  255,
  70,
  140,
  210,
  255,
  50,
  100,
  150,
  255,
]
`;
